CMakeLists.txt

```diff
@@ -24,35 +24,373 @@
+        OpenGL (vdb_view), Boost::Thread (vdb_view), OpenEXR (vdb_render)
+        IlmBase (vdb_render)
+
+    OpenVDB Python
+      Main CMake: openvdb/python/CMakeLists.txt
+      Required: (OpenVDB Core + Dependencies), Python, Boost::Python
+
+    OpenVDB Unit Tests
+      Main CMake: openvdb/unittest/CMakeLists.txt
+      Required: (OpenVDB Core + Dependencies), CppUnit
+
+  ########################################################################
+
+  The CMake supports building the various components of OpenVDB against
+  a prior installation of OpenVDB.
+
+#]=======================================================================]
 
 PROJECT ( OpenVDB )
+CMAKE_MINIMUM_REQUIRED ( VERSION 3.3 )
+# Monitoring <PackageName>_ROOT variables
+IF ( POLICY CMP0048 )
+  CMAKE_POLICY ( SET CMP0048 NEW )
+ENDIF ()
 
-ENABLE_TESTING()
+###### OpenVDB Build/Component Options
 
-find_program(CCACHE_FOUND ccache)
-if(CCACHE_FOUND)
-    set_property(GLOBAL PROPERTY RULE_LAUNCH_COMPILE ccache)
-    set_property(GLOBAL PROPERTY RULE_LAUNCH_LINK ccache)
-endif(CCACHE_FOUND)
+INCLUDE ( CMakeDependentOption )
 
+# @todo SSE/AVX instruction options
+# @todo static/shared library options
+OPTION ( OPENVDB_BUILD_CORE "Build the OpenVDB core" ON )
+OPTION ( OPENVDB_BUILD_BINARIES "Build the vdb binaries" ON )
+OPTION ( OPENVDB_BUILD_PYTHON_MODULE "Build the pyopenvdb Python module" ON )
 OPTION ( OPENVDB_BUILD_UNITTESTS "Build the OpenVDB unit tests" ON )
 OPTION ( OPENVDB_BUILD_DOCS "Build the OpenVDB documentation" OFF )
-OPTION ( OPENVDB_BUILD_PYTHON_MODULE "Build the pyopenvdb Python module" ON )
 OPTION ( OPENVDB_BUILD_HOUDINI_SOPS "Build the Houdini SOPs" OFF )
 OPTION ( OPENVDB_BUILD_MAYA_PLUGIN "Build the Maya plugin" OFF )
-OPTION ( OPENVDB_BUILD_CORE "Build the OpenVDB core" ON )
 OPTION ( OPENVDB_ENABLE_RPATH "Build with RPATH information" ON )
-OPTION ( OPENVDB_ENABLE_3_ABI_COMPATIBLE "Build with OpenVDB v3 ABI" ON )
-OPTION ( OPENVDB_DISABLE_BOOST_IMPLICIT_LINKING "Disable the implicit linking of Boost libraries on Windows" ON )
 OPTION ( OPENVDB_CXX_STRICT "Enable or disable a set of pre-defined compiler warnings for clang and gcc" OFF )
+OPTION ( OPENVDB_ENABLE_3_ABI_COMPATIBLE "Build with OpenVDB ABI 3 (deprecated - see OPENVDB_ABI_VERSION_NUMBER)" OFF )
+
+###### Dependency options
+
+# @todo scalable/concurrent malloc option(s)
+# @todo non-required behavior for blosc, log4cplus (i.e. use if available)
+# @todo OPENVDB_USE_NUMPY options for NumPy
+OPTION ( USE_HOUDINI [=[
+Build the library against a Houdini installation. Turns on autmatically if OPENVDB_BUILD_HOUDINI_SOPS is enabled.
+When enabled, you do not need to provide dependency locations for OpenEXR, IlmBase, TBB, Boost and Blosc. You can
+optionally provide your own version of Boost when building against Houdini 16.5 and gerater.]=] OFF )
+OPTION ( USE_BLOSC "Build the library with blosc support." ON )
+OPTION ( USE_LOG4CPLUS "Build the library with log4cplus support." OFF )
```

Comment:
When these options are ON, cmake will enforce a required dependency to the given lib, erroring if it's not found. Blosc is (currently) on by default, the others are not. Although we recommend people use things like blosc, it's a bit weird to have an optional dependency fail the build for most users. Another way would be to remove all these options completely and bring these deps in if they are found. This would mean that the build wouldn't error if these are missing, unless we provided additional options/messages to do so. In prod I want the build to fail in these explicit cases but that's probably undesirable generally. So my question becomes, whats a good cmake workflow for optional dependencies? I almost want three options; auto, on, off. Or perhaps they should all just be off by default.

Reply:
I agree, it's undesirable to fail the build due to a missing *optional* dependency.  That's part of the reason it's, erm, *optional*.

"Use if found" (via `_FOUND`) is a good option, but can be confusing if the user explicitly doesn't want the option (ie, it's OFF, but used because it was found?).

To enforce dependencies in production, perhaps another option (such as `-D OPENVDB_PRODUCTION_BUILD=ON` can override *all* *optional* dependencies to ON, making them required?  That way there's no question whether any are missed -- a single switch covers all.

Reply:
I think I'm actually leaning to have them all switched off by default - We actually don't use log4cplus, one of the optional deps, in production so a blanket "use all deps" wouldn't immediately help (though arguably we should be using log4cplus :| ). I'm curious to see how other projects handle stuff like enabling components based on available deps e.g. turning a module on or off if a dependency is found. Does embree do anything like this?

Reply:
I've turned off BLOSC for now to match all other optional deps. I still like the idea of using it if it's available but not confident on the way I'd implement this